internal/component/database_observability/mysql/collector/setup_consumers.go

```diff
@@ -0,0 +1,141 @@
+package collector
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.uber.org/atomic"
+
+	"github.com/grafana/alloy/internal/runtime/logging/level"
+)
+
+const (
+	SetupConsumersName = "setup_consumers"
+)
+
+type SetupConsumerArguments struct {
+	DB       *sql.DB
+	Registry *prometheus.Registry
+
+	Logger         log.Logger
+	ScrapeInterval time.Duration
+}
+
+type SetupConsumers struct {
+	dbConnection         *sql.DB
+	Registry             *prometheus.Registry
+	collectInterval      time.Duration
+	SetupConsumersMetric *prometheus.GaugeVec
+
+	logger  log.Logger
+	running *atomic.Bool
+	ctx     context.Context
+	cancel  context.CancelFunc
+}
+
+func NewSetupConsumer(args SetupConsumerArguments) (*SetupConsumers, error) {
+	setupConsumerMetric := prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "database_observability",
+		Name:      "setup_consumers_enabled",
+		Help:      "Whether each performance_schema consumer is enabled (1) or disabled (0)",
+	}, []string{"consumer_name"})
+
+	args.Registry.MustRegister(setupConsumerMetric)
+
+	return &SetupConsumers{
+		dbConnection:         args.DB,
+		Registry:             args.Registry,
+		SetupConsumersMetric: setupConsumerMetric,
+		running:              &atomic.Bool{},
+		logger:               args.Logger,
+		collectInterval:      5 * time.Second,
+	}, nil
+}
+
+func (c *SetupConsumers) Name() string {
+	return SetupConsumersName
+}
+
+func (c *SetupConsumers) Start(ctx context.Context) error {
+	level.Debug(c.logger).Log("msg", SetupConsumersName+" collector started")
+	c.running.Store(true)
+
+	ctx, cancel := context.WithCancel(ctx)
+	c.ctx = ctx
+	c.cancel = cancel
+
+	go func() {
```

Comment:
I think we can avoid spinning up a goroutine here by implementing the prom `Collector` interface. The "prepared statements" collector in the POC app is an example of that approach.

Reason is that by registering a Collector with the Registry passed in, the fetch logic will be run automatically at each Alloy scrape.

Reply:
Doesn't that mean that we can no longer respect the collect scrape interval?

Reply:
Synced with Matt, I think we are not going to do this. Resolving for now, let me know if I undestood.

Reply:
Synced with Matt, I think we are not going to do this. Resolving for now, let me know if I misunderstood.

Reply:
Synced with Matt, I think we are not going to do this. Resolving for now, let me know if I misunderstood.